File: typescript/packages/node/src/resource/registry.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { normalizeFields, type Resource } from '@struktoai/mirage-core'

/**
 * Construct a resource by registry name. Mirrors Python's
 * `mirage.resource.registry.build_resource`.
 *
 * Each entry is an async factory that lazy-imports its module so that
 * importing this file doesn't pull in every backend's dependencies.
 * Only the resource you actually request gets loaded — important for
 * S3/Redis whose peer deps (`@aws-sdk/client-s3`, `redis`) are optional.
 *
 * Configs are normalized from Python-style snake_case (used in YAML and
 * by the Python `mirage.config` loader) to TS-idiomatic camelCase. So
 * the same YAML file works in both Python and TS.
 */
export type ResourceFactory = (config: Record<string, unknown>) => Promise<Resource>

const REGISTRY: Record<string, ResourceFactory> = {
  ram: async (_config) => {
    const { RAMResource } = await import('@struktoai/mirage-core')
    return new RAMResource()
  },
  disk: async (config) => {
    const { DiskResource } = await import('./disk/disk.ts')
    const norm = normalizeFields(config) as { root: string }
    return new DiskResource(norm)
  },
  redis: async (config) => {
    const { RedisResource } = await import('./redis/redis.ts')
    const norm = normalizeFields(config)
    return new RedisResource(norm)
  },
  s3: async (config) => {
    const { S3Resource } = await import('./s3/s3.ts')
    const { normalizeS3Config } = await import('./s3/config.ts')
    return new S3Resource(normalizeS3Config(config))
  },
  gcs: async (config) => {
    const { GCSResource } = await import('./gcs/gcs.ts')
    const { normalizeGcsConfig } = await import('./gcs/config.ts')
    return new GCSResource(normalizeGcsConfig(config))
  },
  oci: async (config) => {
    const { OCIResource } = await import('./oci/oci.ts')
    const { normalizeOciConfig } = await import('./oci/config.ts')
    return new OCIResource(normalizeOciConfig(config))
  },
  r2: async (config) => {
    const { R2Resource } = await import('./r2/r2.ts')
    const { normalizeR2Config } = await import('./r2/config.ts')
    return new R2Resource(normalizeR2Config(config))
  },
  supabase: async (config) => {
    const { SupabaseResource } = await import('./supabase/supabase.ts')
    const { normalizeSupabaseConfig } = await import('./supabase/config.ts')
    return new SupabaseResource(normalizeSupabaseConfig(config))
  },
  databricks_volume: async (config) => {
    const { DatabricksVolumeResource } = await import('./databricks_volume/databricks_volume.ts')
    const { normalizeDatabricksVolumeConfig } = await import('@struktoai/mirage-core')
    return DatabricksVolumeResource.create(normalizeDatabricksVolumeConfig(config))
  },
  postgres: async (config) => {
    const { PostgresResource } = await import('./postgres/postgres.ts')
    const { normalizePostgresConfig } = await import('@struktoai/mirage-core')
    return new PostgresResource(normalizePostgresConfig(config))
  },
  mongodb: async (config) => {
    const { MongoDBResource } = await import('./mongodb/mongodb.ts')
    const { normalizeMongoDBConfig } = await import('@struktoai/mirage-core')
    return new MongoDBResource(normalizeMongoDBConfig(config))
  },
  slack: async (config) => {
    const { SlackResource } = await import('./slack/slack.ts')
    const { normalizeSlackConfig } = await import('./slack/config.ts')
    return new SlackResource(normalizeSlackConfig(config))
  },
  ssh: async (config) => {
    const { SSHResource } = await import('./ssh/ssh.ts')
    const { normalizeSshConfig } = await import('./ssh/config.ts')
    return new SSHResource(normalizeSshConfig(config))
  },
  discord: async (config) => {
    const { DiscordResource } = await import('./discord/discord.ts')
    const { normalizeDiscordConfig } = await import('./discord/config.ts')
    return new DiscordResource(normalizeDiscordConfig(config))
  },
  trello: async (config) => {
    const { TrelloResource } = await import('./trello/trello.ts')
    const { normalizeTrelloConfig } = await import('./trello/config.ts')
    return new TrelloResource(normalizeTrelloConfig(config))
  },
  linear: async (config) => {
    const { LinearResource } = await import('./linear/linear.ts')
    const { normalizeLinearConfig } = await import('./linear/config.ts')
    return new LinearResource(normalizeLinearConfig(config))
  },
  notion: async (config) => {
    const { NotionResource } = await import('./notion/notion.ts')
    const { normalizeNotionConfig } = await import('./notion/config.ts')
    return new NotionResource(normalizeNotionConfig(config))
  },
  langfuse: async (config) => {
    const { LangfuseResource } = await import('./langfuse/langfuse.ts')
    const { normalizeLangfuseConfig } = await import('./langfuse/config.ts')
    return new LangfuseResource(normalizeLangfuseConfig(config))
  },
  github: async (config) => {
    const { GitHubResource } = await import('./github/github.ts')
    const { normalizeGitHubConfig } = await import('./github/config.ts')
    return GitHubResource.create(normalizeGitHubConfig(config))
  },
  github_ci: async (config) => {
    const { GitHubCIResource } = await import('./github_ci/github_ci.ts')
    const { normalizeGitHubCIConfig } = await import('./github_ci/config.ts')
    return new GitHubCIResource(normalizeGitHubCIConfig(config))
  },
  gdocs: async (config) => {
    const { GDocsResource } = await import('./gdocs/gdocs.ts')
    const { normalizeGDocsConfig } = await import('./gdocs/config.ts')
    return new GDocsResource(normalizeGDocsConfig(config))
  },
  gsheets: async (config) => {
    const { GSheetsResource } = await import('./gsheets/gsheets.ts')
    const { normalizeGSheetsConfig } = await import('./gsheets/config.ts')
    return new GSheetsResource(normalizeGSheetsConfig(config))
  },
  gslides: async (config) => {
    const { GSlidesResource } = await import('./gslides/gslides.ts')
    const { normalizeGSlidesConfig } = await import('./gslides/config.ts')
    return new GSlidesResource(normalizeGSlidesConfig(config))
  },
  gdrive: async (config) => {
    const { GDriveResource } = await import('./gdrive/gdrive.ts')
    const { normalizeGDriveConfig } = await import('./gdrive/config.ts')
    return new GDriveResource(normalizeGDriveConfig(config))
  },
  dropbox: async (config) => {
    const { DropboxResource } = await import('./dropbox/dropbox.ts')
    const { normalizeDropboxConfig } = await import('./dropbox/config.ts')
    return new DropboxResource(normalizeDropboxConfig(config))
  },
  box: async (config) => {
    const { BoxResource } = await import('./box/box.ts')
    const { normalizeBoxConfig } = await import('./box/config.ts')
    return new BoxResource(normalizeBoxConfig(config))
  },
  gmail: async (config) => {
    const { GmailResource } = await import('./gmail/gmail.ts')
    const { normalizeGmailConfig } = await import('./gmail/config.ts')
    return new GmailResource(normalizeGmailConfig(config))
  },
  email: async (config) => {
    const { EmailResource } = await import('./email/email.ts')
    const { normalizeEmailConfig } = await import('./email/config.ts')
    return new EmailResource(normalizeEmailConfig(config))
  },
}

/**
 * Look up the registered names. Lazily-mutated by `register()` so users
 * can extend the registry with custom resources.
 */
export function knownResources(): string[] {
  return Object.keys(REGISTRY).sort()
}

/**
 * Register a custom resource factory under `name`. Existing entries
 * are overwritten, mirroring Python's mutable REGISTRY dict.
 */
export function register(name: string, factory: ResourceFactory): void {
  REGISTRY[name] = factory
}

/**
 * Build a resource instance by registry name. Throws if the name is
 * unknown.
 */
export async function buildResource(
  name: string,
  config: Record<string, unknown> = {},
): Promise<Resource> {
  const factory = REGISTRY[name]
  if (factory === undefined) {
    throw new Error(
      `unknown resource ${JSON.stringify(name)}; known: ${knownResources().join(', ')}`,
    )
  }
  return factory(config)
}

File: python/mirage/resource/registry.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from typing import TYPE_CHECKING, NamedTuple

from mirage.resource.loader import load_backend_class

if TYPE_CHECKING:
    from mirage.resource.base import BaseResource


class ResourceEntry(NamedTuple):
    resource_path: str
    config_path: str | None


REGISTRY: dict[str, ResourceEntry] = {
    "ram":
    ResourceEntry("mirage.resource.ram:RAMResource", None),
    "disk":
    ResourceEntry("mirage.resource.disk:DiskResource", None),
    "redis":
    ResourceEntry("mirage.resource.redis:RedisResource", None),
    "s3":
    ResourceEntry("mirage.resource.s3:S3Resource",
                  "mirage.resource.s3:S3Config"),
    "r2":
    ResourceEntry("mirage.resource.r2:R2Resource",
                  "mirage.resource.r2:R2Config"),
    "oci":
    ResourceEntry("mirage.resource.oci:OCIResource",
                  "mirage.resource.oci:OCIConfig"),
    "supabase":
    ResourceEntry("mirage.resource.supabase:SupabaseResource",
                  "mirage.resource.supabase:SupabaseConfig"),
    "gcs":
    ResourceEntry("mirage.resource.gcs:GCSResource",
                  "mirage.resource.gcs:GCSConfig"),
    "minio":
    ResourceEntry("mirage.resource.minio:MinIOResource",
                  "mirage.resource.minio:MinIOConfig"),
    "ceph":
    ResourceEntry("mirage.resource.ceph:CephResource",
                  "mirage.resource.ceph:CephConfig"),
    "wasabi":
    ResourceEntry("mirage.resource.wasabi:WasabiResource",
                  "mirage.resource.wasabi:WasabiConfig"),
    "backblaze":
    ResourceEntry("mirage.resource.backblaze:BackblazeResource",
                  "mirage.resource.backblaze:BackblazeConfig"),
    "digitalocean":
    ResourceEntry("mirage.resource.digitalocean:DigitalOceanResource",
                  "mirage.resource.digitalocean:DigitalOceanConfig"),
    "tencent":
    ResourceEntry("mirage.resource.tencent:TencentResource",
                  "mirage.resource.tencent:TencentConfig"),
    "aliyun":
    ResourceEntry("mirage.resource.aliyun:AliyunResource",
                  "mirage.resource.aliyun:AliyunConfig"),
    "scaleway":
    ResourceEntry("mirage.resource.scaleway:ScalewayResource",
                  "mirage.resource.scaleway:ScalewayConfig"),
    "qingstor":
    ResourceEntry("mirage.resource.qingstor:QingStorResource",
                  "mirage.resource.qingstor:QingStorConfig"),
    "hf_buckets":
    ResourceEntry("mirage.resource.hf_buckets:HfBucketsResource",
                  "mirage.resource.hf_buckets:HfBucketsConfig"),
    "hf_datasets":
    ResourceEntry("mirage.resource.hf_datasets:HfDatasetsResource",
                  "mirage.resource.hf_datasets:HfDatasetsConfig"),
    "hf_models":
    ResourceEntry("mirage.resource.hf_models:HfModelsResource",
                  "mirage.resource.hf_models:HfModelsConfig"),
    "hf_spaces":
    ResourceEntry("mirage.resource.hf_spaces:HfSpacesResource",
                  "mirage.resource.hf_spaces:HfSpacesConfig"),
    "github":
    ResourceEntry("mirage.resource.github:GitHubResource",
                  "mirage.resource.github:GitHubConfig"),
    "github_ci":
    ResourceEntry("mirage.resource.github_ci:GitHubCIResource",
                  "mirage.resource.github_ci:GitHubCIConfig"),
    "linear":
    ResourceEntry("mirage.resource.linear:LinearResource",
                  "mirage.resource.linear:LinearConfig"),
    "gdocs":
    ResourceEntry("mirage.resource.gdocs:GDocsResource",
                  "mirage.resource.gdocs:GDocsConfig"),
    "gsheets":
    ResourceEntry("mirage.resource.gsheets:GSheetsResource",
                  "mirage.resource.gsheets:GSheetsConfig"),
    "gslides":
    ResourceEntry("mirage.resource.gslides:GSlidesResource",
                  "mirage.resource.gslides:GSlidesConfig"),
    "gdrive":
    ResourceEntry("mirage.resource.gdrive:GoogleDriveResource",
                  "mirage.resource.gdrive:GoogleDriveConfig"),
    "slack":
    ResourceEntry("mirage.resource.slack:SlackResource",
                  "mirage.resource.slack:SlackConfig"),
    "discord":
    ResourceEntry("mirage.resource.discord:DiscordResource",
                  "mirage.resource.discord:DiscordConfig"),
    "gmail":
    ResourceEntry("mirage.resource.gmail:GmailResource",
                  "mirage.resource.gmail:GmailConfig"),
    "trello":
    ResourceEntry("mirage.resource.trello:TrelloResource",
                  "mirage.resource.trello:TrelloConfig"),
    "mongodb":
    ResourceEntry("mirage.resource.mongodb:MongoDBResource",
                  "mirage.resource.mongodb:MongoDBConfig"),
    "postgres":
    ResourceEntry("mirage.resource.postgres:PostgresResource",
                  "mirage.resource.postgres:PostgresConfig"),
    "notion":
    ResourceEntry("mirage.resource.notion:NotionResource",
                  "mirage.resource.notion:NotionConfig"),
    "langfuse":
    ResourceEntry("mirage.resource.langfuse:LangfuseResource",
                  "mirage.resource.langfuse:LangfuseConfig"),
    "ssh":
    ResourceEntry("mirage.resource.ssh:SSHResource",
                  "mirage.resource.ssh:SSHConfig"),
    "email":
    ResourceEntry("mirage.resource.email:EmailResource",
                  "mirage.resource.email:EmailConfig"),
    "dify":
    ResourceEntry("mirage.resource.dify:DifyResource",
                  "mirage.resource.dify:DifyConfig"),
    "chroma":
    ResourceEntry("mirage.resource.chroma:ChromaResource",
                  "mirage.resource.chroma:ChromaConfig"),
    "databricks_volume":
    ResourceEntry("mirage.resource.databricks_volume:DatabricksVolumeResource",
                  "mirage.resource.databricks_volume:DatabricksVolumeConfig"),
    "nextcloud":
    ResourceEntry("mirage.resource.nextcloud:NextcloudResource",
                  "mirage.resource.nextcloud:NextcloudConfig"),
    "lancedb":
    ResourceEntry("mirage.resource.lancedb:LanceDBResource",
                  "mirage.resource.lancedb:LanceDBConfig"),
}


def build_resource(name: str, config: dict | None = None) -> "BaseResource":
    """Construct a resource instance by its registry name.

    Resolves resource and config classes lazily via importlib, so
    importing this module does not pull in every resource's
    dependencies. Only the resources actually used get loaded.

    Args:
        name (str): registry key such as ``"s3"`` or ``"ram"``.
        config (dict | None): kwargs for the resource's ``Config``
            class when one exists; otherwise raw resource kwargs
            (e.g. ``{"root": "/tmp"}`` for ``"disk"``).

    Returns:
        BaseResource: a fresh resource instance.

    Raises:
        KeyError: ``name`` is not in ``REGISTRY``.
    """
    if name not in REGISTRY:
        raise KeyError(f"unknown resource {name!r}; known: {sorted(REGISTRY)}")
    entry = REGISTRY[name]
    resource_cls = load_backend_class(entry.resource_path)
    cfg_dict = dict(config or {})
    if entry.config_path is None:
        return resource_cls(**cfg_dict)
    config_cls = load_backend_class(entry.config_path)
    return resource_cls(config_cls(**cfg_dict))
